moped-editor/src/views/projects/projectView/ToolbarPlugin.js

```diff
@@ -0,0 +1,360 @@
+    icon: <Redo />,
+    label: "Redo",
+  },
+  {
+    id: richTextAction.clear,
+    icon: <FormatClear />,
+    label: "Clear formatting",
+  },
+];
+
+const ToolbarPlugin = ({ noteAddSuccess, classes }) => {
+  const [editor] = useLexicalComposerContext();
+
+  const [disableMap, setDisableMap] = useState({
+    [richTextAction.undo]: true,
+    [richTextAction.redo]: true
+  });
+  const [selectionMap, setSelectionMap] = useState({});
+
+  const checkListType = (selection, listType) => {
+    let hasListType = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((selectedNode) => {
+      const selectedNodeParent = selectedNode.getParent()
+      const selectedNodeGrandparent = selectedNodeParent?.getParent();
+      const selectedNodeGreatGrandparent = selectedNodeGrandparent?.getParent();
+      // Check for list formatting and confirm the list type
+      if ($isListNode(selectedNodeParent)) {
+        hasListType = selectedNodeParent.__listType === listType;
+      }
+      // Check for a list item and confirm the parent list type
+      if ($isListItemNode(selectedNodeParent)) {
+        hasListType = selectedNodeGrandparent.__listType === listType;
+      }
+      // to do: iterate through higher level nodes rather than do manual checks like this
+      if ($isListItemNode(selectedNodeGrandparent)) {
+        hasListType = selectedNodeGreatGrandparent.__listType === listType;
+      }
+    })
+    return hasListType;
+  };
+
+  // Checks for selected list formatting, removes if already applied or adds if not applied
+  const handleListUpdate = (command, listType) => {
+    const isListApplied = selectionMap[listType];
+    isListApplied ?
+      editor.dispatchCommand(REMOVE_LIST_COMMAND) :
+      editor.dispatchCommand(command);
+  };
+
+  const checkLink = (selection) => {
+    let isLink = false;
+    const selectedNodes = selection.getNodes();
+    selectedNodes.forEach((selectedNode) => {
+      const selectedNodeParent = selectedNode.getParent()
+      if ($isLinkNode(selectedNodeParent)) {
+        isLink = true
+      }
+    })
+    return isLink;
+  };
+
+  // to do: add url validation and figure out why links are not opening in a new tab
+  const handleLinkUpdate = () => {
+    const isLinkApplied = selectionMap["link"];
+    isLinkApplied ?
+      editor.dispatchCommand(TOGGLE_LINK_COMMAND, null)
+      :
+      navigator.clipboard.readText().then(val => {
+        editor.dispatchCommand(TOGGLE_LINK_COMMAND, {
+          url: val, target: "_blank"
+        })
+      }).catch(err => {
+        console.error('Failed to read clipboard contents: ', err);
+      })
+  };
+
+  const clearFormatting = () => {
+    const formatTypes = Object.keys(selectionMap);
+    formatTypes.forEach((formatType) => {
+      if (selectionMap[formatType]) {
+        onAction(formatType);
+      }
+    });
+  };
+
+  useEffect(() => {
+    // to do: move this function outside of useEffect without causing warnings
```

Comment:
Just wanted to update that this makes more sense to me now that I read https://lexical.dev/docs/concepts/listeners and https://lexical.dev/docs/concepts/commands. I also found this code example which looks like provided some good inspiration. https://github.com/ozanyurtsever/verbum/blob/master/src/plugins/ToolbarPlugin/ToolbarPlugin.tsx. I think it would be awesome to point to some links in this code since it is new to all of us. 📚